src/plugins/timeseries-plugin.ts

```diff
@@ -0,0 +1,270 @@
+                    {
+                        text: `Timestamp: ${round(data.timestamp)}ms`,
+                    },
+                ],
+                this.interactionsEngine.getGlobalMouse()
+            );
+            return true;
+        }
+        return false;
+    }
+
+    private normalizeValue(value: number, heightPerValueUnit: number) {
+        return this.height - value * heightPerValueUnit;
+    }
+
+    private convertDataPointToPointWithIndex(
+        idx: number,
+        timestamp: number,
+        value: number,
+        heightPerValueUnit: number
+    ): TimeseriesPointWithExtra {
+        return {
+            idx,
+            timestamp,
+            value,
+            position: this.renderEngine.timeToPosition(timestamp),
+            normalizedValue: this.normalizeValue(value, heightPerValueUnit),
+        };
+    }
+
+    override render() {
+        if (this.data?.length === 0 || !this.summary) {
+            return;
+        }
+
+        const timestampStart = this.renderEngine.positionX;
+        const timestampEnd = this.renderEngine.positionX + this.renderEngine.getRealView();
+
+        const positionStart = this.renderEngine.timeToPosition(timestampStart);
+        const positionEnd = this.renderEngine.timeToPosition(timestampEnd);
+
+        const heightPerValueUnit =
+            (this.height - this.styles.padding) / Math.max(this.summary.max - this.summary.min, 1);
+
+        let indexStart = 0;
+
+        this.renderEngine.setCtxColor(this.styles.color);
+        this.renderEngine.ctx.beginPath();
+        this.renderEngine.ctx.moveTo(positionStart, this.height);
+
+        let foundStart: boolean = false;
+        let prevTPI: TimeseriesPointWithExtra = EmptyTimeseriesPoint;
+        let tpi: TimeseriesPointWithExtra = EmptyTimeseriesPoint;
+
+        let idx = 0;
+        for (const [time, value] of this.data) {
+            if (time < timestampStart) {
+                indexStart = idx;
+            }
+
+            if (time >= timestampStart && time <= timestampEnd) {
+                if (!foundStart) {
+                    foundStart = true;
+
+                    prevTPI = this.convertDataPointToPointWithIndex(
+                        indexStart,
+                        this.data[indexStart][0],
+                        this.data[indexStart][1],
+                        heightPerValueUnit
+                    );
+                    this.renderEngine.ctx.lineTo(positionStart, prevTPI.normalizedValue);
+                }
+
+                tpi = this.convertDataPointToPointWithIndex(indexStart, time, value, heightPerValueUnit);
+
+                if (prevTPI.idx > -1) {
+                    if (this.styles.lineStyle === 'step') {
+                        this.renderEngine.ctx.lineTo(tpi.position, prevTPI.normalizedValue);
+                    }
+
+                    this.renderEngine.ctx.lineTo(tpi.position, tpi.normalizedValue);
+
+                    this.interactionsEngine.addHitRegion(
```

Comment:
You can avoid creating hit regions and just calculate the plot value from the mouse position directly in the tooltip's render function. Use a binary search to find the specific nearest point: https://en.wikipedia.org/wiki/Binary_search_algorithm

It should look something like this: https://stackoverflow.com/questions/64737346/binary-search-to-find-closest-float-value-to-target-number

Reply:
Why is it important:
If there are 10,000 points on the chart, then the engine will need to brute-force check all 10,000 regions for mouse entry. And most importantly, these 10,000 hit areas will be created on every frame. In the case of a binary search, only 13 simple comparison operations are required (base 2 logarithm of 10000)

Reply:
I think i understand what you are saying; But i am not sure which apis to use on the render; 

Please can you describe clearer how you would expect this work.... here is my guess
 - respond to the mouse-move intereaction event; use the mouse.positionX binary search closest data point and set this on the plugin
 - create 1 hit region for the whole of the chart?
 - use the renderToolip() method?
 
 I had a quick  play - but i am not see the renderTooltip method called frequently with this approach

I also tried rendering the tooltip from the move event ... but this had obvious rendering issues

![image](https://user-images.githubusercontent.com/1444266/213258452-22474645-efb2-48a4-b32f-61144de622be.png)

Reply:
You can only use the `renderTooltip` method and return true from this method. Inside it, you have access to the interactions engine and you can get the position of the mouse from it. I think it should look something like this
```ts
override renderTooltip() {
  const posX = this.interactionsEngine.mouse.x;
  const { time, value } = this.findNearestPoint(pos.x);

  this.renderEngine.renderTooltipFromData(
      [{ text: time }, { text: value }],
      this.interactionsEngine.getGlobalMouse()
  );

  return true;
}
```
The engine should take care of all the work, you just need to describe what exactly the tooltip should show inside the `renderTooltip` method.

If something does not go according to plan, then I think using your code I can figure it out and make changes to the engine, if necessary.)

Reply:
renderTooltip isn't being fired often enough; It does fire - but rarely. Thus the tooltip is stale often. I just put a console.log() inside it to see when it was called.

Reply:
I checked the engine code better - sorry, but I forgot that you really need to mark the hit region.  I think that for this plugin, you need to make only one region for the entire height and width of the chart.

Reply:
Yup - this does improve things - but it isn't perfect. - I have asked other questions in PR